### DISCUSSION TITLE
Publishing to github pages with quarto

### DISCUSSION THREAD

**Discussion by njtierney** [quarto-dev/quarto-cli#7776] [Q&A]
### Description

Hi there!

I'm trying to create a minimal reprex on my issues with publishing to quarto via gh pages. 

My goal is that I want the book to be rebuilt upon pushes to main. I don't want to push HTML content onto the site. I don't want to have to run commands from the command line every time I want to publish to github (although it is cool that that works!)

I see that this is an [issue which has been discussed](https://github.com/quarto-dev/quarto-cli/issues/5686), but I can't actually seem to find an answer to my problem.

My process - following on from https://quarto.org/docs/publishing/github-pages.html#publish-command

1. via rstudio, create a quarto book, "quarto-gh-pages"
2. create github repo, ["quarto-gh-pages" ](https://github.com/njtierney/quarto-gh-pages)
3. push content onto repo - [first commit here](https://github.com/njtierney/quarto-gh-pages/commit/041d570ef7f650dcdd1e2527891fe0a0f0d54e60)
4. run `quarto publish gh-pages` in the command line of the repo:

<img width="581" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/6488485/a5dd0061-9ed0-4fd4-9795-d3f963cb2ed4">

This then produces a book as a website! Woo!

<img width="1222" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/6488485/c1622423-d544-4b91-8ac0-428e6ad639a5">

But my goal is that I want the book to be rebuilt upon pushes to main. I don't want to push HTML content onto the site.

When I push another change to the repo, no new build. See [this commit](https://github.com/njtierney/quarto-gh-pages/commit/70b52e9f738d98ad26f6b0ed05d88fabb59f3558)

<img width="1248" alt="image" src="https://github.com/quarto-dev/quarto-cli/assets/6488485/d07d0e82-181b-48fe-83fd-7e1c55d59857">

Follow instructions from quarto website:

> Once you’ve completed a local publish, add a publish.yml GitHub Action to your project by creating this YAML file and saving it to .github/workflows/publish.yml:

So I add a [publish.yml file](https://github.com/njtierney/quarto-gh-pages/commit/0f8183385e2c445b8be75020327a0ced3851bf34)

[This then fails](https://github.com/njtierney/quarto-gh-pages/actions/runs/7084015388/job/19277456791#step:13:69): 

> Error in loadNamespace(x) : there is no package called ‘rmarkdown’

OK, so I need to have R installed and R packages available now? Perhaps add renv?

Following instructions from: https://rstudio.github.io/renv/articles/ci.html

I add [R and renv in this commit](https://github.com/njtierney/quarto-gh-pages/commit/a1ec8be187b4dc811305189c5d871920b4f8873c)

[Error due to no lockfile](https://github.com/njtierney/quarto-gh-pages/actions/runs/7084070596/job/19277621022)

> Error: This project does not contain a lockfile.

[OK, add a lockfile](https://github.com/njtierney/quarto-gh-pages/commit/4df4e64fa6fcaced834c9458aedc4de5c893d6fb)

[But TeX isn't installed](https://github.com/njtierney/quarto-gh-pages/actions/runs/7084115342/job/19277755479#step:6:151):

> No TeX installation was detected.
> Please run 'quarto install tinytex' to install TinyTex.
> If you prefer, you may install TexLive or another TeX distribution.

Alright, so, [add tinytex](https://github.com/njtierney/quarto-gh-pages/commit/4c822cfc0488a8bf047931474c9584eb76b16eaa)

[This worked!](https://github.com/njtierney/quarto-gh-pages/commits/main)

This seems like a fair number of hoops to need to run through, I'm wondering if there's a way to automate a few of these steps? Am I missing something? It seems like I'm probably missing something, but I'd really appreciate some guidance here, as I feel like there's some great opportunities to make this as easy as possible.

**Comment by mcanouil** [comment]
As you noticed, we are aware of the "publish to GitHub Pages" documentation issue which is a bit outdated and the fact that current default error about missing `_publish.yml` this is actually only for Netlify and means "you need to run `quarto publish gh-pages` locally first.

If you don't use PDF as output, there is no need for TinyTex, if you don't have computations you don't need either Jupyter or R.
It's not really feasible to guess any user dependencies to be set in GitHub Actions, that's up to the user unfortunately.

Note that you have several examples and documentation of GitHub Actions workflows at <https://github.com/quarto-dev/quarto-actions>.
This link is provided in Continuous integration section about GitHub Actions: <https://quarto.org/docs/publishing/ci.html#github-actions>

**Comment by njtierney** [comment]
Thanks for the quick response!

And I should also preface this by saying firstly thank you everyone for writing such a useful piece of software, thanks!

> As you noticed, we are aware of the "publish to GitHub Pages" documentation issue which is a bit outdated and the fact that current default error about missing _publish.yml this is actually only for Netlify and means "you need to run quarto publish gh-pages locally first.

If I understand correctly, `quarto publish gh-pages` only needs to be run when using netlify? So this section: https://quarto.org/docs/publishing/github-pages.html#publish-action is inaccurate now?

What happens if the user doesn't use `quarto publish gh-pages`? Can they still use a github action and it will publish to github pages for them? Or is it that the user needs to create a gh-pages branch, and that quarto command does that?

> If you don't use PDF as output, there is no need for TinyTex, if you don't have computations you don't need either Jupyter or R.
It's not really feasible to guess any user dependencies to be set in GitHub Actions, that's up to the user unfortunately.

Yup, that's totally fair!

**Reply by mcanouil** [reply]
Netlify is another hosting services which requires the _publish.yml.

GitHub Pages does not need that.
But currently, if you set your GitHub workflow without first publishing locally, the workflow will fail showing an error involving _publish.yml. This is a currently wrong error message.
When you use `quarto publish gh-pages` you are publishing to GitHub Pages, nowhere else.

As I said and as stated in the documentation (that part is correct), no matter if you want to use GitHub Action or not: you first need to use `quarto publish gh-pages` locally. It will do everything except setting GitHub action workflow.

**Comment by johann-wagner** [comment]
Not too sure if this helps, but I found this video useful:
[Publish a Quarto project using GitHub Pages+GitHub Actions in 6 minutes (no need to render locally!)](https://www.youtube.com/watch?v=arzBRW5XIkg)
